zilliqa/src/pool.rs

```diff
@@ -67,75 +116,105 @@ impl TransactionPool {
     /// consecutive with a previously returned transaction, from the same sender.
     pub fn best_transaction(&mut self) -> Option<VerifiedTransaction> {
         loop {
-            let Some(ReadyItem {
-                from_addr, nonce, ..
-            }) = self.ready.pop()
-            else {
+            let Some(ReadyItem { tx_index, .. }) = self.ready.pop() else {
                 return None;
             };
-            let Some(transaction) = self.transactions.remove(&(from_addr, nonce)) else {
-                // A transaction might have been ready, but then we learnt that the sender's nonce increased, making it
-                // invalid. In this case, we will have removed the transaction from `transactions` in `update_nonce`.
+            let Some(transaction) = self.transactions.remove(&tx_index) else {
+                // A transaction might have been ready, but it might have gotten popped
+                // or the sender's nonce might have increased, making it invalid. In this case,
+                // we will have removed the transaction from `transactions` (in `update_nonce`
+                // or `pop_transaction`), but a stale reference would still exist in the heap.
+                //
                 // We loop until we find a transaction that hasn't been made invalid.
                 continue;
             };
-            self.hash_to_index.remove(&transaction.hash);
 
-            if let Some(next_txn) = self.transactions.get(&(from_addr, nonce + 1)) {
-                self.ready.push(next_txn.into());
+            // If we've popped a nonced transaction, that may have made a subsequent one valid
+            if let TxIndex::Nonced(address, nonce) = tx_index {
```

Comment:
I might be overthinking it, but a slightly cleaner way could be something like
```rust
impl TxIndex {
  fn next(&self) -> Option<TxIndex> {
    match self {
      TxIndex::Nonced(address, nonce) => Some(TxIndex::Nonced(address, nonce + 1)),
      _ => None,
    }
  }
}
```
and using `if let Some(next) = tx_index.next() {` here?

Reply:
I like this!